tests/test_websockets.py

```diff
@@ -0,0 +1,242 @@
+
+        with fake_websocket_broker.serve(response):
+            with pytest.raises(WebsocketConnectionError) as exc:
+                mqttc.connect("localhost", 1888, keepalive=10)
+
+        assert str(exc.value) == "WebSocket handshake error, invalid secret key"
+
+
+@pytest.mark.parametrize("proto_ver,proto_name", [
+    (client.MQTTv31, "MQIsdp"),
+    (client.MQTTv311, "MQTT"),
+])
+class TestValidHeaders(object):
+    """ Testing for functionality in request/response headers """
+
+    def _get_callback_handler(self, response_headers, check_request=None):
+        """ Get a basic BaseRequestHandler which returns the information in
+        self._response_headers
+        """
+
+        class WebsocketHandler(socketserver.BaseRequestHandler):
+            def handle(_self):
+                self.data = _self.request.recv(1024).strip()
+                print("Received '{:s}'".format(self.data.decode("utf8")))
+
+                decoded = self.data.decode("utf8")
+
+                if check_request is not None:
+                    check_request(decoded)
+
+                # Create server hash
+                GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
+                key = re.search("sec-websocket-key: ([A-Za-z0-9+/=]*)", decoded, re.IGNORECASE).group(1)
+
+                to_hash = "{:s}{:s}".format(key, GUID)
+                hashed = hashlib.sha1(to_hash.encode("utf8"))
+                encoded = base64.b64encode(hashed.digest()).decode("utf8")
+
+                response_headers["Sec-WebSocket-Accept"] = encoded
+
+                # Respond with the correct hash
+                response = get_websocket_response(response_headers)
+
+                _self.request.sendall(response)
+
+        return WebsocketHandler
+
+    def test_successful_connection(self, proto_ver, proto_name,
+                                   fake_websocket_broker,
+                                   init_response_headers):
+        """ Connect successfully, on correct path """
+
+        mqttc = client.Client(
+            "test_successful_connection",
+            protocol=proto_ver,
+            transport="websockets"
+            )
+
+        response = self._get_callback_handler(init_response_headers)
+
+        with fake_websocket_broker.serve(response):
+            mqttc.connect("localhost", 1888, keepalive=10)
+
+            mqttc.disconnect()
+
+    @pytest.mark.parametrize("mqtt_path", [
+        "/mqtt"
+        "/special",
+        None,
+    ])
+    def test_correct_path(self, proto_ver, proto_name, fake_websocket_broker,
+                          mqtt_path, init_response_headers):
+        """ Make sure it can connect on user specified paths """
+
+        mqttc = client.Client(
+            "test_correct_path",
+            protocol=proto_ver,
+            transport="websockets"
+            )
+
+        mqttc.ws_set_options(
+            path=mqtt_path,
+        )
+
+        def create_response_hash(decoded):
```

Comment:
This method is never used. I think it should be the "check_request" parameter of _get_callback_handler